contrib/ros/envs/sawyer_env.py

```diff
@@ -0,0 +1,346 @@
+from rllab.envs.base import Step
+from rllab.misc import logger
+from rllab.spaces import Box
+
+from contrib.ros.envs.ros_env import RosEnv
+from contrib.ros.robots.sawyer import Sawyer
+from contrib.ros.util.common import rate_limited
+from contrib.ros.util.gazebo import Gazebo
+
+
+def _goal_distance(goal_a, goal_b):
+    """
+    :param goal_a:
+    :param goal_b:
+    :return distance: distance between goal_a and goal_b
+    """
+    assert goal_a.shape == goal_b.shape
+    return np.linalg.norm(goal_a - goal_b, axis=-1)
+
+
+class SawyerEnv(RosEnv, Serializable):
+    """
+    ROS Sawyer Env
+    """
+
+    def __init__(self,
+                 initial_robot_joint_pos,
+                 robot_control_mode,
+                 has_object,
+                 distance_threshold,
+                 initial_model_pos,
+                 sparse_reward,
+                 target_in_the_air,
+                 simulated=False,
+                 target_range=0.15,
+                 obj_range=0.15):
+        """
+        :param initial_robot_joint_pos: {str: float}
+                {'joint_name': value}
+        :param robot control mode: str
+                'effort'/'position'/'velocity'
+        :param has_object: bool
+                if there is object in this experiment
+        :param distance_threshold: float
+                When the distance from robot endpoint to the target point is smaller than this threshold
+                the training episode is done.
+                To calculate the sparse reward, if the distance is smaller than this threshold returns -1
+                else returns 0.
+        :param initial_model_pos: dict
+                joint positions and object positions
+        :param sparse_reward: bool
+                if reward_type is 'sparse'
+        :param target_in_the_air: bool
+                set to see if the target can be in the air
+        :param simulated: bool
+                if the environment is for real robot or simulation
+        :param target_range: float
+                the new target position for the new training episode would be randomly sampled inside
+                [initial_gripper_pos - target_range, initial_gripper_pos + target_range]
+        :param obj_range: float
+                the new initial object x,y position for the new training episode would be randomly sampled inside
+                [initial_gripper_pos - obj_range, initial_gripper_pos + obj_range]
+        """
+        # run the superclass initialize function first
+        Serializable.quick_init(self, locals())
+
+        RosEnv.__init__(self)
+        # long-running ROS/robot setup
+        self.initialize()
+
+        # Verify robot is enabled
+        self._robot = Sawyer(
+            initial_robot_joint_pos, control_mode=robot_control_mode)
+        if not self._robot.enabled:
+            raise RuntimeError(
+                '[Error] SawyerEnv.__init__: The robot is not enabled!')
```

Comment:
Remove the "[Error] SawyerEnv.__init__:'"stuff. (1) It will be incorrect if this code is ever refactored and (2) all of that information is provided by Python when a RuntimeError is thrown.

Reply:
Does the robot need to be enabled for the constructor to finish? why?